File: AardvarkSample/AardvarkSample/SampleViewController.m
//
//  SampleViewController.m
//  AardvarkSample
//
//  Created by Dan Federman on 10/11/14.
//  Copyright 2014 Square, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import <Aardvark/Aardvark.h>
#import <CoreAardvark/CoreAardvark.h>

#import "SampleAppDelegate.h"
#import "SampleViewController.h"


NSString *const SampleViewControllerTapLogKey = @"SampleViewControllerTapLog";


@interface SampleViewController ()

@property (nonatomic) ARKLogStore *tapGestureLogStore;
@property (nonatomic) UITapGestureRecognizer *tapRecognizer;

@end


@implementation SampleViewController

#pragma mark - UIViewController

- (void)viewDidLoad;
{
    [super viewDidLoad];
    
    self.tapGestureLogStore = [[ARKLogStore alloc] initWithPersistedLogFileName:@"SampleTapLogs.data"];
    self.tapGestureLogStore.name = @"Taps";
    
    // Ensure that the tap log store will only store tap logs.
    self.tapGestureLogStore.logFilterBlock = ^(ARKLogMessage *logMessage) {
        return [logMessage.userInfo[SampleViewControllerTapLogKey] boolValue];
    };
    
    // Do not log tap logs to the main tap log store.
    [ARKLogDistributor defaultDistributor].defaultLogStore.logFilterBlock = ^(ARKLogMessage *logMessage) {
        return (BOOL)![logMessage.userInfo[SampleViewControllerTapLogKey] boolValue];
    };
    
    [[ARKLogDistributor defaultDistributor] addLogObserver:self.tapGestureLogStore];
    
    ARKEmailBugReporter *bugReporter = ((SampleAppDelegate *)[UIApplication sharedApplication].delegate).bugReporter;
    [bugReporter addLogStores:@[self.tapGestureLogStore]];
}

- (void)viewDidAppear:(BOOL)animated;
{
    ARKLog(@"%s", __PRETTY_FUNCTION__);
    
    self.tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(_tapDetected:)];
    self.tapRecognizer.cancelsTouchesInView = NO;
    [self.view addGestureRecognizer:self.tapRecognizer];
    
    [super viewDidAppear:animated];
}

- (void)viewDidDisappear:(BOOL)animated;
{
    ARKLog(@"%s", __PRETTY_FUNCTION__);
    
    [self.tapRecognizer.view removeGestureRecognizer:self.tapRecognizer];
    self.tapRecognizer = nil;
    
    [super viewDidDisappear:animated];
}

#pragma mark - Actions

- (IBAction)viewARKLogMessages:(id)sender;
{
    ARKLog(@"%s", __PRETTY_FUNCTION__);
    ARKLogTableViewController *defaultLogsViewController = [ARKLogTableViewController new];
    [self.navigationController pushViewController:defaultLogsViewController animated:YES];
}

- (IBAction)viewTapLogs:(id)sender;
{
    ARKLog(@"%s", __PRETTY_FUNCTION__);
    ARKLogTableViewController *tapLogsViewController = [[ARKLogTableViewController alloc] initWithLogStore:self.tapGestureLogStore logFormatter:[ARKDefaultLogFormatter new]];
    [self.navigationController pushViewController:tapLogsViewController animated:YES];
}

- (IBAction)blueButtonPressed:(id)sender;
{
    ARKLog(@"Blue");
}

- (IBAction)redButtonPressed:(id)sender;
{
    ARKLog(@"Red");
}

- (IBAction)greenButtonPressed:(id)sender;
{
    ARKLog(@"Green");
}

- (IBAction)yellowButtonPressed:(id)sender;
{
    ARKLog(@"Yellow");
}

#pragma mark - Private Methods

- (void)_tapDetected:(UITapGestureRecognizer *)tapRecognizer;
{
    if (tapRecognizer == self.tapRecognizer && tapRecognizer.state == UIGestureRecognizerStateEnded) {
        ARKLogWithType(ARKLogTypeDefault, @{ SampleViewControllerTapLogKey : @YES }, @"Tapped %@", NSStringFromCGPoint([tapRecognizer locationInView:nil]));
    }
}

@end

File: AardvarkSample/AardvarkSample/SampleViewController.swift
//
//  SampleViewController.swift
//  AardvarkSample
//
//  Created by Dan Federman on 9/21/16.
//  Copyright © 2016 Square, Inc. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import Aardvark
import CoreAardvark
import UIKit


class SampleViewController : UIViewController {
    
    let tapLogKey: NSString = "SampleViewcontrollerTapLog"
    
    private var tapRecognizer: UITapGestureRecognizer?
    private var tapGestureLogStore: ARKLogStore?
    
    // MARK: – Initialization
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(tapDetected(tapRecognizer:)))
    }
    
    // MARK: – UIViewController
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        guard let tapGestureLogStore = ARKLogStore(persistedLogFileName: "SampleTapLogs.data") else {
            return
        }
        
        tapGestureLogStore.name = "Taps"
        
        // Ensure that the tap log store will only store tap logs.
        tapGestureLogStore.logFilterBlock = { [weak self] logMessage in
            guard let `self` = self else {
                return false
            }
            
            guard let isTapLog = logMessage.userInfo[self.tapLogKey] as? Bool else {
                return false
            }
            
            return isTapLog
        }
        
        // Do not log tap logs to the main tap log store.
        ARKLogDistributor.default().defaultLogStore.logFilterBlock = { [weak self] logMessage in
            guard let `self` = self else {
                return true
            }
            
            guard let isTapLog = logMessage.userInfo[self.tapLogKey] as? Bool else {
                return true
            }
            
            return !isTapLog
        }
        
        ARKLogDistributor.default().add(tapGestureLogStore)
        
        // Store the log store.
        self.tapGestureLogStore = tapGestureLogStore
        
        guard let bugReporter = (UIApplication.shared.delegate as? SampleAppDelegate)?.bugReporter else {
            return
        }
        
        // Add our log store to the bug reporter.
        bugReporter.add([tapGestureLogStore])
    }
    
    override func viewDidAppear(_ animated: Bool) {
        log("\(#function)")
        
        guard let tapRecognizer = tapRecognizer else {
            return
        }
        
        view.addGestureRecognizer(tapRecognizer)
        
        super.viewDidAppear(animated)
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        log("\(#function)")
        
        guard let tapRecognizer = tapRecognizer else {
            return
        }
        
        tapRecognizer.view?.removeGestureRecognizer(tapRecognizer)
        
        super.viewDidAppear(animated)
    }
    
    // MARK: – Actions
    
    @IBAction func viewARKLogMessages(_ sender: AnyObject) {
        log("\(#function)")
        let defaultLogsViewController = ARKLogTableViewController()
        navigationController?.pushViewController(defaultLogsViewController, animated: true)
    }
    
    @IBAction func viewTapLogs(_ sender: AnyObject) {
        log("\(#function)")
        
        guard let tapGestureLogStore = tapGestureLogStore else {
            return
        }
        
        guard let tapLogsViewController = ARKLogTableViewController(logStore: tapGestureLogStore, logFormatter: ARKDefaultLogFormatter()) else {
            return
        }
        
        navigationController?.pushViewController(tapLogsViewController, animated: true)
    }
    
    @IBAction func blueButtonPressed(_ sender: AnyObject) {
        log("Blue")
    }
    
    @IBAction func redButtonPressed(_ sender: AnyObject) {
        log("Red")
    }
    
    @IBAction func greenButtonPressed(_ sender: AnyObject) {
        log("Green")
    }
    
    @IBAction func yellowButtonPressed(_ sender: AnyObject) {
        log("Yellow")
    }
    
    // MARK: – Private Methods
    
    @objc
    private func tapDetected(tapRecognizer: UITapGestureRecognizer) {
        guard tapRecognizer == self.tapRecognizer && tapRecognizer.state == .ended else {
            return
        }
        
        log("Tapped \(NSStringFromCGPoint(tapRecognizer.location(in: nil)))", userInfo: [tapLogKey : true as NSNumber])
    }
}
